Failed to take snapshot because RocksDb is shutting down

**Describe the bug**
Taking a snapshot in middle of shutting down resulted in: `io.camunda.zeebe.db.ZeebeDbException: Failed to take snapshot in path /usr/local/zeebe/data/raft-partition/partitions/1/pending/270685-24-270663-270664.`. We can prevent this by looking at the exception's status and ignore it if it's due to a shutdown being in progress.

**Log/Stacktrace**
https://console.cloud.google.com/errors/CNON5Mv1u8qCiQE?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911

<details><summary>Full Stacktrace</summary>
 <p>

```
io.camunda.zeebe.db.ZeebeDbException: Failed to take snapshot in path /usr/local/zeebe/data/raft-partition/partitions/1/pending/270685-24-270663-270664.
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.createSnapshot(ZeebeTransactionDb.java:131) ~[zeebe-db-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl.lambda$takeSnapshot$0(StateControllerImpl.java:178) ~[zeebe-broker-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.takeInternal(FileBasedTransientSnapshot.java:67) ~[zeebe-snapshots-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.lambda$take$0(FileBasedTransientSnapshot.java:53) ~[zeebe-snapshots-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:94) [zeebe-util-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.0.0.jar:1.0.0]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:191) [zeebe-util-1.0.0.jar:1.0.0]
Caused by: org.rocksdb.RocksDBException: ShutdownInProgress
	at org.rocksdb.Checkpoint.createCheckpoint(Native Method) ~[rocksdbjni-6.19.3.jar:?]
	at org.rocksdb.Checkpoint.createCheckpoint(Checkpoint.java:51) ~[rocksdbjni-6.19.3.jar:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.createSnapshot(ZeebeTransactionDb.java:128) ~[zeebe-db-1.0.0.jar:1.0.0]
	... 9 more
```

</p>
</details>


**Environment:**
 - OS: <!-- [e.g. Linux] -->
 - Zeebe Version: <!-- [e.g. 0.20.0] -->
 - Configuration: <!-- [e.g. exporters etc.] -->

--- a/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java
+++ b/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/StateControllerImpl.java
@@ -35,33 +35,37 @@ public class StateControllerImpl implements StateController, PersistedSnapshotLi
   private static final ReplicationContext INVALID_SNAPSHOT = new ReplicationContext(null, -1, null);
   private static final Logger LOG = Loggers.SNAPSHOT_LOGGER;
 
-  private final int partitionId;
-
   private final SnapshotReplication replication;
   private final Map<String, ReplicationContext> receivedSnapshots =
       new Object2NullableObjectHashMap<>();
 
   private final Path runtimeDirectory;
+
+  @SuppressWarnings("rawtypes")
   private final ZeebeDbFactory zeebeDbFactory;
+
+  @SuppressWarnings("rawtypes")
   private final ToLongFunction<ZeebeDb> exporterPositionSupplier;
+
   private final AtomixRecordEntrySupplier entrySupplier;
 
   private final SnapshotReplicationMetrics metrics;
 
+  @SuppressWarnings("rawtypes")
   private ZeebeDb db;
+
   private final ConstructableSnapshotStore constructableSnapshotStore;
   private final ReceivableSnapshotStore receivableSnapshotStore;
 
   public StateControllerImpl(
       final int partitionId,
-      final ZeebeDbFactory zeebeDbFactory,
+      @SuppressWarnings("rawtypes") final ZeebeDbFactory zeebeDbFactory,
       final ConstructableSnapshotStore constructableSnapshotStore,
       final ReceivableSnapshotStore receivableSnapshotStore,
       final Path runtimeDirectory,
       final SnapshotReplication replication,
       final AtomixRecordEntrySupplier entrySupplier,
-      final ToLongFunction<ZeebeDb> exporterPositionSupplier) {
-    this.partitionId = partitionId;
+      @SuppressWarnings("rawtypes") final ToLongFunction<ZeebeDb> exporterPositionSupplier) {
     this.constructableSnapshotStore = constructableSnapshotStore;
     this.receivableSnapshotStore = receivableSnapshotStore;
     this.runtimeDirectory = runtimeDirectory;
@@ -101,18 +105,15 @@ public class StateControllerImpl implements StateController, PersistedSnapshotLi
             snapshotIndexedEntry.term(),
             lowerBoundSnapshotPosition,
             exportedPosition);
+
     transientSnapshot.ifPresent(this::takeSnapshot);
+
     return transientSnapshot;
   }
 
   @Override
   public void consumeReplicatedSnapshots() {
     replication.consume(this::consumeSnapshotChunk);
-  }
-
-  @Override
-  public void stopConsumeReplicatedSnapshots() {
-    replication.stopConsuming();
   }
 
   @Override
@@ -143,6 +144,7 @@ public class StateControllerImpl implements StateController, PersistedSnapshotLi
   }
 
   @Override
+  @SuppressWarnings("rawtypes")
   public ZeebeDb openDb() {
     if (db == null) {
       db = zeebeDbFactory.createDb(runtimeDirectory.toFile());
@@ -158,10 +160,17 @@ public class StateControllerImpl implements StateController, PersistedSnapshotLi
   }
 
   @Override
+  public void stopConsumeReplicatedSnapshots() {
+    replication.stopConsuming();
+  }
+
+  @Override
   public void close() throws Exception {
     if (db != null) {
-      db.close();
+      final var dbToClose = db;
       db = null;
+      dbToClose.close();
+
       LOG.debug("Closed database from '{}'.", runtimeDirectory);
     }
 

